lib/maths/CBoostedTreeLeafNodeStatistics.cc

```diff
@@ -325,68 +275,134 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
 
     SSplitStatistics result;
 
-    for (auto i : featureBag) {
-        std::size_t c{m_MissingDerivatives[i].s_Count};
-        double g{m_MissingDerivatives[i].s_Gradient};
-        double h{m_MissingDerivatives[i].s_Curvature};
-        for (const auto& derivatives : m_Derivatives[i]) {
-            c += derivatives.s_Count;
-            g += derivatives.s_Gradient;
-            h += derivatives.s_Curvature;
+    // We seek to find the value at the minimum of the quadratic expansion of the
+    // regularized loss function. For a given leaf this expansion is
+    //
+    //   L(w) = 1/2 w^t H(\lambda) w + g^t w
+    //
+    // where H(\lambda) = \sum_i H_i + \lambda I, g = \sum_i g_i and w is the leaf's
+    // weight. Here, g_i and H_i denote an example's loss gradient and Hessian and i
+    // ranges over the examples in the leaf. Completing the square and noting that
+    // the minimum is given when the quadratic form is zero we have
```

Comment:
The last sentence is unclear: what does it mean to complete the square? And I guess, the derivative of the quadratic form should be zero.

Reply:
I added an extended comment and made the condition that H is positive definite clear. See [3009f42](https://github.com/elastic/ml-cpp/pull/993/commits/3009f4252f9b52adf2284b9cc06dca738ba57ab7).